### ISSUE TITLE
[apex] Inner Type Order

### ISSUE THREAD

**Issue by mitchspano** [pmd/pmd#5430]
**Proposed Rule Name:** [apex] Inner Type Order

**Proposed Category:** Code Style

**Description:** This rule enforces a consistent ordering of inner types within Apex classes. It specifies that methods should be declared first, followed by inner enums, then inner interfaces, and finally inner classes. This promotes code readability and maintainability by providing a predictable structure for developers to follow.

**Code Sample:**

```java
// Violates the rule: Incorrect order of inner types
public class MyClass {

    // Inner class before inner enum
    public class InnerClass { 
        // ...
    }

    public enum InnerEnum {
        // ...
    }

    public void myMethod() { 
        // ...
    }
}

// Compliant code: Correct order of inner types
public class MyClass {

    public void myMethod() { 
        // ...
    }

    public enum InnerEnum {
        // ...
    }

    public interface InnerInterface {
        // ...
    }

    public class InnerClass { 
        // ...
    }
}
```

**Possible Properties:**

*  N/A

**Comment by sirephil** [comment]
I much prefer placing nested classes before methods, so that types are introduced before their use. Methods are always last for me.

**Comment by pkozuchowski** [comment]
Not sure if I'd use this rule, because it would cause more false flags than anything. If I have 2 interfaces + implementations in a class then I want to see them in order Interface, Implementation A, Implementation B. Sometimes enums first - methods after will be the right order for better readability.

**Comment by Damecek** [comment]
Some form of ordering should definitely be in place, but it will be challenging to agree on a **single** approach that everyone can accept.
